pkg/psmdb/mongos.go

```diff
@@ -0,0 +1,361 @@
+			)
+		}
+	}
+
+	if msSpec.AuditLog != nil && msSpec.AuditLog.Destination == api.AuditLogDestinationFile {
+		if msSpec.AuditLog.Filter == "" {
+			msSpec.AuditLog.Filter = "{}"
+		}
+		args = append(args,
+			"--auditDestination=file",
+			"--auditFilter="+msSpec.AuditLog.Filter,
+			"--auditFormat="+string(msSpec.AuditLog.Format),
+		)
+		switch msSpec.AuditLog.Format {
+		case api.AuditLogFormatBSON:
+			args = append(args, "--auditPath="+MongodContainerDataDir+"/auditLog.bson")
+		default:
+			args = append(args, "--auditPath="+MongodContainerDataDir+"/auditLog.json")
+		}
+	}
+
+	return args
+}
+
+func volumes(cr *api.PerconaServerMongoDB) []corev1.Volume {
+	fvar := false
+	t := true
+	volumes := []corev1.Volume{
+		{
+			Name: InternalKey(cr),
+			VolumeSource: corev1.VolumeSource{
+				Secret: &corev1.SecretVolumeSource{
+					DefaultMode: &secretFileMode,
+					SecretName:  InternalKey(cr),
+					Optional:    &fvar,
+				},
+			},
+		},
+		{
+			Name: "ssl",
+			VolumeSource: corev1.VolumeSource{
+				Secret: &corev1.SecretVolumeSource{
+					SecretName:  cr.Spec.Secrets.SSL,
+					Optional:    &cr.Spec.UnsafeConf,
+					DefaultMode: &secretFileMode,
+				},
+			},
+		},
+		{
+			Name: "ssl-internal",
+			VolumeSource: corev1.VolumeSource{
+				Secret: &corev1.SecretVolumeSource{
+					SecretName:  cr.Spec.Secrets.SSLInternal,
+					Optional:    &t,
+					DefaultMode: &secretFileMode,
+				},
+			},
+		},
+		{
+			Name: MongodDataVolClaimName,
+			VolumeSource: corev1.VolumeSource{
+				EmptyDir: &corev1.EmptyDirVolumeSource{},
+			},
+		},
+	}
+
+	if *cr.Spec.Mongod.Security.EnableEncryption {
+		volumes = append(volumes,
+			corev1.Volume{
+				Name: cr.Spec.Mongod.Security.EncryptionKeySecret,
+				VolumeSource: corev1.VolumeSource{
+					Secret: &corev1.SecretVolumeSource{
+						DefaultMode: &secretFileMode,
+						SecretName:  cr.Spec.Mongod.Security.EncryptionKeySecret,
+						Optional:    &fvar,
+					},
+				},
+			},
+		)
+	}
+
+	return volumes
+}
+
+func MongosService(cr *api.PerconaServerMongoDB) corev1.Service {
+	svc := corev1.Service{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: "v1",
+			Kind:       "Service",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      cr.Name + "-" + "mongos",
+			Namespace: cr.Namespace,
+		},
+	}
+
+	if cr.Spec.Sharding.Mongos != nil {
+		svc.Annotations = cr.Spec.Sharding.Mongos.ServiceAnnotations
+	}
+
+	return svc
+}
+
+func MongosServiceSpec(cr *api.PerconaServerMongoDB) corev1.ServiceSpec {
+	ls := map[string]string{
+		"app.kubernetes.io/component": "mongos",
+	}
+
+	spec := corev1.ServiceSpec{
+		Ports: []corev1.ServicePort{
+			{
+				Name:       mongosPortName,
+				Port:       cr.Spec.Sharding.Mongos.Port,
+				TargetPort: intstr.FromInt(int(cr.Spec.Sharding.Mongos.Port)),
+			},
+		},
+		Selector:                 ls,
+		LoadBalancerSourceRanges: cr.Spec.Sharding.Mongos.LoadBalancerSourceRanges,
+	}
+
+	if !cr.Spec.Sharding.Mongos.Expose.Enabled {
+		spec.ClusterIP = "None"
```

Comment:
hm, why we don't set it if cr.Spec.Sharding.Mongos.Expose.Enabled==true ?

Reply:
https://github.com/percona/percona-server-mongodb-operator/pull/439/commits/1f82100b29e90a9cc32f66ec5829533d8875b06b